enhancements/cloud-controller-manager/openstack-cloud-controller-manager.md

```diff
@@ -0,0 +1,173 @@
+
+Actions:
+
+- Create a new repo in OpenShift’s github: https://github.com/openshift/cluster-cloud-controller-manager-operator
+
+- Implement the operator, using [library-go](https://github.com/openshift/library-go) primitives.
+
+#### Build the operator image by OCP automation
+
+To start using CCM operator in OpenShift we need to build its image and make sure it is a part of the OpenShift release image. The image should be automatically tested before it becomes available. Dockerfile should be a part of the operator's repo.
+
+Actions:
+
+- Configure CI operator to build CCM operator image.
+
+CI operator will run containerized and End-to-End tests and also push the resulting image in the OCP Quay account.
+
+#### Integrate the solution with OpenShift
+
+Actions:
+
+- Make sure Cinder CSI driver is supported in OpenShift.
+
+- Add CCM operator support in Cluster Version Operator. CCM operator will be deployed on early stages of installation and then it deploys and configures OpenStack CCM itself.
+
+- Change Kubelet configuration for OpenStack in [Machine Config Operator templates](https://github.com/openshift/machine-config-operator/blob/master/templates/master/01-master-kubelet/openstack/units/kubelet.service.yaml#L30) to adopt external cloud provider.
+
+- Change Config observer in [library-go](https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go#L154) in the same manner. Then the library should be bumped in `cluster-kube-apiserver-operator` and `cluster-kube-controller-manager-operator`.
+
+### Risks and Mitigations
+
+#### OpenStack CCM doesn’t work properly on OpenShift
+
+The manager has not been tested on either OSP or OCP. The team uses Devstack + K8s plugin for the development.
+
+Severity: medium-high
+Likelihood: probably
+
+#### OpenShift components can't work with the CCM properly
+
+Since OpenShift imposes additional limitations compared to Kubernetes, some collisions are possible.
+So far there are no CCMs available in OpenShift, and there can be some problems with operability.
+
+Severity: medium-high
+Likelihood: low
+
+#### Cinder CSI driver is not available in OpenShift
+
+Integration with OpenStack Cinder (Persistent Volume Manager) was included in the in-tree cloud provider. But CCM doesn't contain this feature, so Cinder CSI driver must be installed in the system to work properly. Despite the fact that a cluster with some limitations can be installed without the driver, it will be impossible to work on it in production.
+
+Severity: medium-high
+Likelihood: low
+
+## Design Details
+
+### Test Plan
+
+Testing will consist in running of serial and parallel e2e-openstack jobs on the modified system with OpenStack CCM. If they pass successfully, we can consider the solution works.
+
+No changes to CI system are required, but additional tests should be run on a cloud with self-signed certificates.
+
+### Upgrade / Downgrade Strategy
```

Comment:
Would be nice to add Graduation Criteria. Please detail the expectations for 4.7 and 4.8 releases.

Reply:
Would also be good to have the upgrade process documented, how do existing clusters migrate to the new CCM?

Reply:
done